R/get_analysis_date.R

```diff
@@ -0,0 +1,270 @@
+#'   fail_rate = temp$fail_rate,                                       # failure rate
+#'   dropout_rate = temp$dropout_rate)                                 # dropout rate
+#'
+#' # example 1: cut for analysis at the 24-th month.
+#' get_analysis_date(
+#'   simu_data,
+#'   planned_calendar_time = 24)
+#'
+#' # example 2: cut for analysis when there are 300 events
+#' # in the overall population.
+#' get_analysis_date(
+#'   simu_data,
+#'   target_event_overall = 300)
+#'
+#' # example 3: cut for analysis at the 24-th month,
+#' # and there are 300 events in the overall population,
+#' # which arrives later.
+#' get_analysis_date(
+#'   simu_data,
+#'   planned_calendar_time = 24,
+#'   target_event_overall = 300)
+#'
+#' # example 4: cut for analysis when there are at least 100
+#' # events in the biomarker-positive population,
+#' # and at least 200 events in the biomarker-negative population,
+#' # which arrives later.
+#' get_analysis_date(
+#'   simu_data,
+#'   target_event_per_stratum = c(100, 200))
+#' get_analysis_date(
+#'   simu_data,
+#'   target_event_overall = 150,
+#'   target_event_per_stratum = c(100, NA))
+#'
+#' # example 5: cut for analysis when there are at least 100 events
+#' # in the biomarker positive population,
+#' # and at least 200 events in the biomarker negative population, which arrives later.
+#' # But will stop at the 30-th months if events less than 100/200.
+#' get_analysis_date(
+#'   simu_data,
+#'   target_event_per_stratum = c(100, 200),
+#'   max_extension_for_target_event = 30)
+#'
+#' # example 6: cut for analysis after 12 months followup when 80%
+#' # of the patients are enrolled in the overall population.
+#' get_analysis_date(
+#'   simu_data,
+#'   enroll_rate = enroll_rate,
+#'   min_n_overall = n * 0.8,
+#'   min_followup = 12)
+#'
+#' # example 7: cut for analysis when 12 months after at least
+#' # 200/160 patients are enrolled at the
+#' # biomarker positive/negative population.
+#' get_analysis_date(
+#'   simu_data,
+#'   enroll_rate = enroll_rate,
+#'   min_n_per_stratum = c(200, 160),
+#'   min_followup = 12)
+#' get_analysis_date(
+#'   simu_data,
+#'   enroll_rate = enroll_rate,
+#'   min_n_per_stratum = c(200, NA),
+#'   min_followup = 12)
+get_analysis_date <- function(
+    simu_data,
+    # Option 1: planned calendar time for the analysis
+    planned_calendar_time = NA,
+    # Option 2: reach targeted events
+    target_event_overall = NA,
+    target_event_per_stratum = NA,
+    # Option 3: max time extension to reach targeted events
+    max_extension_for_target_event = NA,
+    # Option 4: planned minimum time after the previous analysis
+    previous_analysis_date = 0,
+    min_time_after_previous_analysis = NA,
+    # Option 5: minimal follow-up time after specified enrollment fraction
+    enroll_rate = NA,
+    min_n_overall = NA,
+    min_n_per_stratum = NA,
+    min_followup = NA
+){
+  # input checking
+  input_check_scale <- function(x = NA, label = "x"){
```

Comment:
Unless there is a compelling reason to do so (eg [function factories](https://adv-r.hadley.nz/function-factories.html)), it is best practice not to nest functions inside of other functions. Nested functions are more difficult to reason about (because of R's [scoping rules](https://adv-r.hadley.nz/environments.html#dynamic-scope)), and more difficult to test and interactively debug (eg you can't run `debug(input_check_scale)`). I recommend moving all helper functions to the bottom of this file.